src/neo/SmartContract/Native/Oracle/OracleContract.cs

```diff
@@ -89,29 +79,36 @@ internal override void Initialize(ApplicationEngine engine)
             engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_RequestId), new StorageItem(BitConverter.GetBytes(0ul)));
         }
 
+        [ContractMethod(0, CallFlags.AllowModifyStates)]
+        private void Finish(ApplicationEngine engine)
+        {
+            Transaction tx = (Transaction)engine.ScriptContainer;
+            OracleResponse response = tx.Attributes.OfType<OracleResponse>().First();
+            engine.OracleResponses?.Add(response.Id);
+            OracleRequest request = GetRequest(engine.Snapshot, response.Id);
+            StackItem userData = BinarySerializer.Deserialize(request.UserData, engine.MaxStackSize, engine.MaxItemSize, engine.ReferenceCounter);
+            engine.CallFromNativeContract(null, request.CallbackContract, request.CallbackMethod, request.Url, userData, response.Success, response.Result);
+        }
+
         protected override void PostPersist(ApplicationEngine engine)
         {
             base.PostPersist(engine);
-            foreach (Transaction tx in engine.Snapshot.PersistingBlock.Transactions)
+            foreach (ulong responseId in engine.OracleResponses)
             {
-                //Filter the response transactions
-                OracleResponse response = tx.Attributes.OfType<OracleResponse>().FirstOrDefault();
```

Comment:
So the real optimization is here?

Reply:
we doen't fetch the transactions only the responses, so if there are no oracle tx, the performance will be the same

Reply:
But fetching the transactions is very fast.

Reply:
This is my tests

Test:

```csharp
[TestMethod]
        public void Benchmark()
        {
            // Init

            TestBlockchain.InitializeMockNeoSystem();
            var block = new Block() { Transactions = new Transaction[500] };

            for (var x = 0; x < block.Transactions.Length; x++)
            {
                block.Transactions[x] = TestUtils.GetTransaction(UInt160.Zero);
            }
            // Without oracle txs

            var snapshot = Blockchain.Singleton.GetSnapshot();
            snapshot.PersistingBlock = block;

            var sw = Stopwatch.StartNew();
            for (int x = 0; x < 1_000; x++)
            {
                var clone = snapshot.Clone();
                Test(clone, block);
            }
            sw.Stop();
            Console.WriteLine("Without oracle tx: " + sw.Elapsed);

            // With one oracle tx

            ulong responseId = 5;
            block.Transactions[5].Attributes = new TransactionAttribute[]
            {
                new OracleResponse()
                {
                     Id = responseId,
                     Result = new byte[0],
                     Success = true
                }
            };

            var request = new OracleRequest
            {
                OriginalTxid = block.Transactions[5].Hash,
                GasForResponse = 0,
                Url = "",
                Filter = "",
                CallbackContract = UInt160.Zero,
                CallbackMethod = "",
                UserData = new byte[] { 0 }
            };
            StorageKey key = new KeyBuilder(NativeContract.Oracle.Id, 7).Add(responseId);
            snapshot.Storages.Add(key, new StorageItem(request));
            key = new KeyBuilder(NativeContract.Oracle.Id, 6).Add(OracleContract.GetUrlHash(request.Url));
            snapshot.Storages.GetAndChange(key, () => new StorageItem(new OracleContract.IdList())).GetInteroperable<OracleContract.IdList>().Add(responseId);

            sw = Stopwatch.StartNew();
            for (int x = 0; x < 1_000; x++)
            {
                var clone = snapshot.Clone();
                Test(clone, block, block.Transactions[5]);
            }
            sw.Stop();
            Console.WriteLine("With one oracle tx: " + sw.Elapsed);
        }

        private void Test(StoreView snapshot, Block block, params Transaction[] oracleTxs)
        {
            var responses = new HashSet<ulong>();
            ApplicationEngine engine;

            foreach (var tx in oracleTxs)
            {
                engine = ApplicationEngine.Create(TriggerType.Application, tx, snapshot, 0, true);
                engine.LoadScript(OracleResponse.FixedScript);
                engine.OracleResponses = responses;
                // FAULT in callback
                if (engine.Execute() != VMState.FAULT) throw new InvalidOperationException();
                engine.Dispose();
            }

            // Execute postPersist

            engine = ApplicationEngine.Create(TriggerType.System, null, snapshot, 0, true);
            engine.OracleResponses = responses;
            engine.LoadScript(Blockchain.postPersistScript);
            if (engine.Execute() != VMState.HALT) throw new InvalidOperationException();
            engine.Dispose();
        }

```

Without this PR:

> Without oracle tx: 00:00:00.1375103 / With one oracle tx: 00:00:00.4162590

With this PR:

> Without oracle tx: 00:00:00.0641516 / With one oracle tx: 00:00:00.3524290

Reply:
No difference 😂

Reply:
without oracle tx it's half time

Reply:
It saves 60 milliseconds per block and saves about 16 hours per million blocks. It doesn't look much different.